backend-rust/src/indexer.rs

```diff
@@ -4221,19 +4225,213 @@ impl PreparedSpecialTransactionOutcomes {
+        // PaydayAccountReward to account 2
+        // PaydayPoolReward to `None`
+        // PaydayAccountReward to account 10
+        // PaydayAccountReward to account 3
+        // Means 5, 6 are receiving rewards from delegating to 1 and 2 receiving rewards
+        // from delegating to 8, and 10, 3 are receiving rewards from passive
+        // delegation.
+        //
+        // `PaydayPoolReward` and `PaydayAccountReward` events only occure in payday
+        // blocks starting in protocol 4.
+        //
+        // The mapping `acc_payday_pool_rewards` associates the `pool_owner` (a
+        // `baker_id` or `None`) to its `PaydayPoolRewards`. The `None` value
+        // tracks passive deleagation.
+        let mut acc_payday_pool_rewards: HashMap<Option<u64>, PaydayPoolRewards> = HashMap::new();
+        let mut last_pool_owner: Option<Option<u64>> = None;
+
+        for event in &self.events {
+            match event {
+                SpecialTransactionOutcome::PaydayPoolReward {
+                    pool_owner,
+                    transaction_fees,
+                    baker_reward,
+                    finalization_reward,
+                } => {
+                    // The pool owner is `None` only if the pool rewards are for the passive
+                    // delegators. There is only one event per
+                    // payday block that rewards the passive delegators.
+                    let last = pool_owner.as_ref().map(|baker_id| baker_id.id.into());
+
+                    let acc_payday_pool_rewards =
+                        acc_payday_pool_rewards.entry(last).or_insert(PaydayPoolRewards::new());
+
+                    acc_payday_pool_rewards.transaction_fees.total_amount =
+                        transaction_fees.micro_ccd;
+                    acc_payday_pool_rewards.block_baking.total_amount = baker_reward.micro_ccd;
+                    acc_payday_pool_rewards.block_finalization.total_amount =
+                        finalization_reward.micro_ccd;
+
+                    last_pool_owner = Some(last);
+                }
+                SpecialTransactionOutcome::PaydayAccountReward {
+                    transaction_fees,
+                    baker_reward,
+                    finalization_reward,
+                    account,
+                } => {
+                    // Collect all rewards from the delegators and associate the rewards to their
+                    // baker pools.
+                    if let Some(pool_owner) = last_pool_owner {
+                        if let Some(baker_id) = pool_owner {
+                            let account_index = sqlx::query_scalar!(
+                                "SELECT index
+                                    FROM accounts
+                                    WHERE address = $1",
+                                account.to_string()
+                            )
+                            .fetch_one(tx.as_mut())
+                            .await?;
```

Comment:
Ideally this mapping from address to index was part of the insert, so we avoid all the chatting

Reply:
I optimized the indexer performance with regards to the payday rewards insertion.